File: services/subscription-service/src/SubscriptionService.Domain/Entities/Subscription.cs
using SubscriptionService.Domain.Enums;

namespace SubscriptionService.Domain.Entities;

public class Subscription
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid UserId { get; set; }
    public Guid PlanId { get; set; }
    public SubscriptionStatus Status { get; set; } = SubscriptionStatus.PendingPayment;
    public DateTime PeriodStart { get; set; }
    public DateTime PeriodEnd { get; set; }
    public bool AutoRenew { get; set; } = true;
    public DateTime? CancelledAt { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    public Plan Plan { get; set; } = null!;
    public List<Payment> Payments { get; set; } = new();
    public List<Invoice> Invoices { get; set; } = new();
}

File: web/src/lib/types/subscription.ts
export interface Plan {
  id: string;
  name: string;
  tier: string;
  priceMonthly: number;
  maxScreens: number;
  maxQuality: string;
  features: string;
  isActive: boolean;
}

export interface Subscription {
  id: string;
  userId: string;
  plan: Plan;
  status: string;
  periodStart: string;
  periodEnd: string;
  autoRenew: boolean;
  cancelledAt: string | null;
  createdAt: string;
}

export interface Invoice {
  id: string;
  subscriptionId: string;
  invoiceNumber: string;
  amount: number;
  currency: string;
  periodStart: string;
  periodEnd: string;
  issuedAt: string;
}

export interface Payment {
  id: string;
  subscriptionId: string;
  amount: number;
  currency: string;
  status: string;
  transactionId: string | null;
  cardLastFour: string;
  createdAt: string;
}

export interface CreateSubscriptionRequest {
  planId: string;
  cardNumber: string;
}

export interface CreateSubscriptionResult {
  subscriptionId: string;
  userId: string;
  planId: string;
  planName: string;
  tier: string;
  status: string;
  periodStart: string;
  periodEnd: string;
  amountCharged: number;
  currency: string;
}

export interface ChangePlanRequest {
  newPlanId: string;
  cardNumber: string;
}

export interface ChangePlanResult {
  subscriptionId: string;
  oldPlanId: string;
  oldPlanName: string;
  newPlanId: string;
  newPlanName: string;
  newTier: string;
  priceDifference: number;
  status: string;
}

export interface CancelSubscriptionResult {
  subscriptionId: string;
  status: string;
  cancelledAt: string;
  periodEnd: string;
}
